HPCC-18593 Create helper library to digitally sign text

RussWhitehead opened the pull request:
Create a general purpose library that can use optional public/private
key files to digitally sign text strings. This will initially be used
to authenticate Dali requests when passwords are no longer available
or change every 30 seconds

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

RussWhitehead committed:
HPCC-18593 Create helper library to digitally sign text

Create a general purpose library that can use optional public/private
key files to digitally sign text strings. This will initially be used
to authenticate Dali requests when passwords are no longer available
or change every 30 seconds

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@mckellyln Please review

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 22b6892bb4a32c283f9b86092c61f2bf56da451f
Build: failed
Number of Error(s): 1
-- Configuring incomplete, errors occurred!
Get ECLWatch build result

RussWhitehead committed:
HPCC-18594 Create helper library to encrypt/decrypt text

Create a general purpose library that can use optional public/private key
files to encrypt and decrypt text strings. This will initially be used
to build the workunit security token

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 22b6892bb4a32c283f9b86092c61f2bf56da451f
Build: failed
Number of Error(s): 1
-- Configuring incomplete, errors occurred!
Get ECLWatch build result

AttilaVamos commented on system/security/CMakeLists.txt:
```diff
@@ -24,6 +24,8 @@ if (USE_ZLIB)
   HPCC_ADD_SUBDIRECTORY (zcrypt)
 endif()
 
+HPCC_ADD_SUBDIRECTORY (digiSign)
```
Typo: the directory is "digisign", not "digiSign". That's why it doesn't build.

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 016011bced3309c83b64b55dccaa96788cfa2418
Build: failed
Number of Error(s): 2
CMake Error: Error in cmake code at
-- Configuring incomplete, errors occurred!
Get ECLWatch build result

AttilaVamos commented on system/security/digisign/CMakeLists.txt:
```diff
@@ -0,0 +1,60 @@
+        jlib
+    )
+
+if (USE_ZLIB)s
```
Typo: ')s' instead of ')'

RussWhitehead committed:
HPCC-18593 Create helper library to digitally sign text

Create a general purpose library that can use optional public/private
key files to digitally sign text strings. This will initially be used
to authenticate Dali requests when passwords are no longer available
or change every 30 seconds

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 30afa48f325abff80d8f194e0cf49ac2d24e359b
Build: failed
Number of Error(s): 7
/mnt/disk1/home/vamosax/smoketest/PR-10543/HPCC-Platform/system/security/digisign/digisign.cpp:207:117: error: invalid conversion from \xe2\x80\x98const unsigned char*\xe2\x80\x99 to \xe2\x80\x98unsigned char*\xe2\x80\x99 [-fpermissive]
/usr/include/openssl/evp.h:634:5: error:   initializing argument 2 of \xe2\x80\x98int EVP_DigestVerifyFinal(EVP_MD_CTX*, unsigned char*, size_t)\xe2\x80\x99 [-fpermissive]
make[2]: *** [system/security/digisign/CMakeFiles/digisign.dir/digisign.cpp.o] Error 1
make[1]: *** [system/security/digisign/CMakeFiles/digisign.dir/all] Error 2
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: c76f07907b1198a040ec9df397ddc744a9b51497
Build: failed
Number of Error(s): 7
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
/mnt/disk1/home/vamosax/smoketest/PR-10543/HPCC-Platform/plugins/javaembed/javaembed.cpp:2081:48: error: expected class-name before \xe2\x80\x98,\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10543/HPCC-Platform/plugins/javaembed/javaembed.cpp:2110:5: error: \xe2\x80\x98IXmlWriterExt\xe2\x80\x99 does not name a type
make[2]: *** [plugins/javaembed/CMakeFiles/javaembed.dir/javaembed.cpp.o] Error 1
make[1]: *** [plugins/javaembed/CMakeFiles/javaembed.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

RussWhitehead commented:
Messed this one up good, closing and will reopen another
